Google My Business: Redirect logged out users to login page

stephanethomas opened the pull request:
This pull request fixes an empty page presented to **logged users** when trying to access the new [`Google My Business` page](https://wordpress.com/google-my-business). Instead, they will now be redirected to the [`Login` page](https://wordpress.com/log-in):
 
Before | After
------ | -----
<img width="624" alt="screenshot" src="https://user-images.githubusercontent.com/594356/37777252-5b4af16e-2de7-11e8-9026-d2a505157c32.png"> | <img width="623" alt="screenshot" src="https://user-images.githubusercontent.com/594356/37777367-a463252e-2de7-11e8-866c-18432fc3ee28.png">

This pull request also updates a label, and two links.

#### Testing instructions
 
1. Run `git checkout add/redirect-logged-out-users-to-login` and start your server
2. Open the [`Google My Business` page](http://calypso.localhost:3000/google-my-business) in an incognito window
3. Assert that you are redirected to the [`Login` page](http://calypso.localhost:3000/log-in)
4. Submit valid credentials, or use a magic link to log in to WordPress.com
5. Assert that you are redirected to the `Google My Business` page

You may also give a try other pages from the Google My Business flow:

* http://calypso.localhost:3000/google-my-business/select-business-type
* http://calypso.localhost:3000/google-my-business/new
* http://calypso.localhost:3000/google-my-business/select-location
* http://calypso.localhost:3000/google-my-business/stats

#### Reviews
 
- [ ] Code
- [ ] Product

seear commented on client/controller/index.web.js:
```diff
@@ -60,6 +61,22 @@ export function redirectLoggedIn( context, next ) {
+
+	if ( ! currentUser ) {
+		page.redirect( login( {
+			isNative: config.isEnabled( 'login/native-login-links' ),
```
👍

Tug commented:
Wait, this is only fixing the `/google-my-business` route, whereas every logged in routes currently have that problem:
- https://wordpress.com/sharing
- https://wordpress.com/stats
- ...

Shouldn't we fix this for every logged in routes?
It's just a matter of redirecting to log in here I believe https://github.com/Automattic/wp-calypso/blob/master/client/boot/common.js#L111

seear commented:
> Shouldn't we fix this for every logged in routes?

This would require some extensive testing. A lot of routes probably were not built with this in mind, see for instance the problems in #22993 when the account and domain don't match.

Tug commented:
> This would require some extensive testing.

Yeah agreed.

About #22993, it looks like it would benefit from this type of feature. I read all the comments there but didn't see any mentions of a problem regarding account and domain. You mean if the initial query is not valid or the user did not connect to the right account? That should probably be handled in the checkout controller right?

Let's merge this if it's causing a real bug in production (it doesn't look like it does) but I don't think it's a good sustainable solution to set `enableLoggedOut` to `true` for logged-in only routes just because we want to redirect to `/log-in`... Looks like a paradox.

seear commented:
> You mean if the initial query is not valid or the user did not connect to the right account? That should probably be handled in the checkout controller right?

Sorry, I should have linked to a [pertinent comment](https://github.com/Automattic/wp-calypso/pull/22993#issuecomment-370786733). The trouble is, site and other data may not yet be loaded at controller time, so any validation is necessarily done by components in an ad-hoc fashion.

The site-selection middleware does do some checking of whether a site is owned by the current account, but again it has to wait for data to become available, leading to redirects once the components have started rendering, further adding to the confusion.

stephanethomas commented:
I agree, setting `enableLoggedOut` to `true` just to be able to display a login form to logged-out users doesn't sound right. Since we want to enable this redirect only for specific pages, what do you think of introducing a new prop in [sections](https://github.com/Automattic/wp-calypso/blob/1fbd28e9d06719dcb60375d9650dc9403d01b30b/client/wordpress-com.js#L7) just for that? The [`loggedOutMiddleware` middleware](https://github.com/Automattic/wp-calypso/blob/master/client/boot/common.js#L83) would then only perform this redirect if a section has that prop.

nylen commented:
Drive-by comment inspired by recent Slack discussion: it seems good to ship a simple approach to this first and test it for a few routes, then later, circle back around to making it work in all the places where it's needed.  That second part will probably take a while and it shouldn't hold up making this work where we really need it the most.

stephanethomas commented:
In the case of Google My Business, this change is super low priority. I'd be happy to wait for the Calypso Framework team to come up with a better technical solution. I'm going to freeze this pull request until then.

/cc @bluefuton who was waiting on this for https://github.com/Automattic/wp-calypso/issues/23623

stephanethomas commented:
A heads-up that I've rebased this pull request, and updated it to use the middleware introduced in https://github.com/Automattic/wp-calypso/pull/22993. This is a temporary solution until we have a more appropriate fix for #23785.

Tug commented:
LGTM 👍
